### ISSUE TITLE
MoE grouped_mm backend silently trains on transposed expert weights when `2*moe_intermediate == hidden_size` (bf16 and bnb-4bit)

### ISSUE THREAD

**Issue by pjordanandrsn** [unslothai/unsloth-zoo#849]
`temporary_patches/moe_utils.py::preprocess_weight` converts transformers-v5 F.linear expert
layout (`gate_up (E, 2I, H)`, `down (E, H, I)`) to grouped_mm layout by shape inspection:

```python
    if proj_type == "gate_up":
        # Want (E, hidden_dim, 2*intermediate).
        if weight.shape[1] == hidden_dim:
            return weight
        else:
            return weight.transpose(-2, -1)
```

When `2*moe_intermediate_size == hidden_size` the check is ambiguous — `weight.shape[1]` equals
`hidden_dim` for the F.linear layout too — and the weight comes back **untransposed**, so the
gate_up matmul runs on transposed weights in every expert and every MoE layer (the analogous
`down` branch has the same ambiguity when `moe_intermediate_size == hidden_size`). Loss stays
finite and training appears to run normally, but outputs are numerically wrong. The
`_WEIGHT_PREPROCESSORS` registry that could disambiguate per-arch has no registrations
(2026.6.7 wheel and main as of 2026-07-02).

## Repro

unsloth 2026.6.9 / unsloth-zoo 2026.6.7 / transformers 5.5.0, torch 2.11.0+cu128 (outside
unsloth's `<2.11` pin — the layout logic is version-independent by code reading), any
grouped_mm-capable GPU:

```python
import unsloth, torch, torch.nn.functional as F
from transformers import Qwen3MoeConfig, Qwen3MoeForCausalLM

cfg = Qwen3MoeConfig(vocab_size=1024, hidden_size=64, intermediate_size=128,
                     moe_intermediate_size=32,   # 2*32 == hidden_size -> square gate_up
                     num_hidden_layers=1, num_attention_heads=4, num_key_value_heads=2,
                     head_dim=16, num_experts=8, num_experts_per_tok=2,
                     decoder_sparse_step=1, initializer_range=0.2)
torch.manual_seed(0)
model = Qwen3MoeForCausalLM(cfg).to("cuda", torch.bfloat16).eval()
experts = model.model.layers[0].mlp.experts

T, K = 64, 2
hs = torch.randn(T, cfg.hidden_size, device="cuda", dtype=torch.bfloat16)
idx = torch.randint(0, 8, (T, K), device="cuda")
w = torch.softmax(torch.rand(T, K, device="cuda", dtype=torch.bfloat16), -1)

out = experts(hs, idx, w).float()                      # zoo grouped_mm backend

ref = torch.zeros(T, cfg.hidden_size, device="cuda")   # stock Qwen3MoeExperts math, fp32
gu, dn = experts.gate_up_proj.float(), experts.down_proj.float()
for t in range(T):
    for k in range(K):
        e = idx[t, k]
        gate, up = F.linear(hs[t].float(), gu[e]).chunk(2, -1)
        ref[t] += w[t, k].float() * F.linear(F.silu(gate) * up, dn[e])

print("rel err:", ((out - ref).abs().max() / ref.abs().max()).item())
# grouped_mm backend (default):        rel err 0.994  (silently wrong)
# UNSLOTH_MOE_BACKEND=native_torch:    rel err 0.005  (bf16 noise, correct)
```

Measured on an RTX A2000: max-abs error vs stock math is **139× beyond** a bf16-precision control
on this bf16 repro, 140.7× on the `load_in_4bit` variant of the same model (the bnb-4bit dequant
path, `forward_moe_backend_bnb4bit`, feeds the same `preprocess_weight`, so 4-bit training is
equally affected). Forcing `preprocess_weight` to transpose-always drops the excess-over-noise
ratio to exactly 1.0 — max-abs error identical to the bf16-precision control — and
`torch._grouped_mm` itself is exact vs a hand loop, so the primitive is fine; the layout guess is
the bug. On pinned torch the same grouped_mm default path serves H100-class GPUs.

## Blast radius

As of today, no official checkpoint I scanned actually reaches the silent path: official
Qwen3-MoE, GLM-4.5, ERNIE-4.5, LFM2-8B-A1B and DeepSeek-V3 configs are all non-square, and
gpt-oss-20b/120b are down-square but safe because transformers stores gpt-oss already in grouped
layout, so no-transpose is correct there. But **allenai/OLMoE-1B-7B-0924 (and -0125) has square
gate_up** (2·1024 == 2048) — it currently crashes earlier for an unrelated reason (#850), and the obvious fix for that crash routes it straight into this silent one. Any
pruned/community config with matching dims hits it silently now.

## Suggested fix

The layout is knowable from the source module class — register per-arch preprocessors (the
registry already exists, unused), or key on a `_unsloth_grouped_mm_format` attribute the way zoo
PR #717 did for `GptOssExperts`. Shape-guessing can't be made safe for square dims.

Workaround for affected users: `UNSLOTH_MOE_BACKEND=native_torch`.

Found while verifying the #4032 fix (which works for its Qwen3-30B-A3B target in my testing) —
happy to test candidate patches; the repro and parity harness are ready to rerun.

**Comment by pjordanandrsn** [comment]
Addendum from reading the sibling backends — the blast surface of the ambiguous shape check is
wider than the grouped_mm/bnb4bit routes in the report:

- **FP8 reaches the same `preprocess_weight` twice.** The `_scaled_grouped_mm` fast path calls it
  on the raw float8 tensor (`processed_weight = preprocess_weight(weight, proj_type, hidden_dim,
  model_type)`, `moe_utils_fp8.py:434`, invoked for both projections at `:505-506`), and the
  dequant fallback swaps bf16 copies into `forward_native_grouped_mm`, which calls it again
  (`moe_utils.py:1059/:1190`). The `model_type` escape hatch doesn't currently change anything:
  it comes from `getattr(self, "_unsloth_model_type", None)` — only the deepseek_v3 and gemma4
  patches set that attribute — and nothing in unsloth or unsloth-zoo ever calls
  `register_weight_preprocessor`, so `_WEIGHT_PREPROCESSORS` is empty and the shape branch always
  runs. A square-dims model hits the same silent corruption under FP8.
- **The `unsloth_triton` backend re-implements the same guess inline** (it never calls
  `preprocess_weight`): `if self.gate_up_proj.shape[-1] == hidden_dim:` at `moe_utils.py:1344`,
  the down analogue at `:1401`, and `intermediate_dim = self.gate_up_proj.shape[1] // 2` at
  `:1312`. Notably it also never checks `_unsloth_grouped_mm_format` — the flag
  `forward_native_moe_loop` uses at `:1511-1514`, where the comment already calls the shape check
  "unsafe when intermediate_dim == hidden_dim". So the hazard is documented in one backend and
  live in three others.
- **For contrast, the MXFP4/gpt-oss path is immune** — it derives layout from config fields
  (`module.intermediate_size` / `module.hidden_size`, `gpt_oss.py:492-515`), not tensor-dim
  comparison. That's the fix shape in miniature.

Practical upshot for whoever picks this up: a fix scoped to `preprocess_weight` covers the
grouped_mm, bnb4bit, and both FP8 routes, but the `unsloth_triton` backend needs its own layout
signal (or to honor `_unsloth_grouped_mm_format`) to close the class. Caveat on evidence: the
above is from reading the shipped 2026.6.7 wheel (unchanged on main as of today) — I don't have
Hopper hardware or the grouped-GEMM Triton lib to execute the FP8/Triton paths, so unlike the
original report these two are code-reading findings, not measured ones.

**Comment by pjordanandrsn** [comment]
Correction to my addendum above, after getting the Triton path to actually run (the grouped-GEMM
lib ships inside unsloth itself — my "can't execute locally" caveat was wrong):

**Measured, `unsloth_triton` is NOT affected by the square case.** On the same square repro
(`2*moe_intermediate == hidden`, bnb-4bit), `UNSLOTH_MOE_BACKEND=unsloth_triton` matches the fp32
reference at exactly the bf16-noise floor (excess-over-noise 1.0) — its inline shape checks
(`moe_utils.py:1344/:1401`) happen to resolve this case correctly, unlike `preprocess_weight`.

The narrower concern stands, but as code-reading only: the Triton path never consults
`_unsloth_grouped_mm_format`, so pre-grouped expert modules (the layout family
`forward_native_moe_loop` special-cases at `:1511-1514`) would pass through untransposed there.
I don't have a model on that path to measure, so weight it accordingly.

Net for a fix: `preprocess_weight` covers grouped_mm / bnb4bit / both FP8 routes; the Triton
backend only needs to honor the format flag for pre-grouped modules — it does not need the
square-dims fix.

**Comment by pjordanandrsn** [comment]
Single-file repro: https://gist.github.com/pjordanandrsn/0f2639aac1057305ccbe12a0a26bee2b — on unsloth 2026.6.9 / zoo 2026.6.7, square-dim experts (2·moe_intermediate == hidden) run untransposed through grouped_mm: re-verified today at 139.2× beyond bf16-noise (rel err 111%), and a transpose-always patch drops it to exactly 1.0 (`UNSLOTH_MOE_BACKEND=native_torch` is exact). @Datta0 this intersects the bnb-MoE work from unslothai/unsloth#4032 — happy to PR the fix if you can confirm the intended grouped_mm weight layout.
